lib/libcheri_caprevoke/libcheri_caprevoke.c

```diff
@@ -0,0 +1,438 @@
+/*
+ * Revocation bitmap manipulation operations.
+ */
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdatomic.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <cheri/cheric.h>
+
+#include <sys/param.h>
+#include <cheri/revoke.h>
+#include <machine/cherireg.h>
+#include <machine/pte.h>
+#include <machine/vmparam.h>
+#include "libcaprevoke.h"
+
+#if defined(__riscv)
+#if defined(__CHERI_PURE_CAPABILITY__)
+#include "asm/riscv_purecap.h"
+#else
+#include "asm/riscv_hybrid.h"
+#endif
+#else
+#error "Need libcheri_caprevoke architecture-specific header"
+#endif
+
+static ptrdiff_t caprev_shadow_nomap_first_word_offset(vaddr_t base);
+static ptrdiff_t caprev_shadow_nomap_last_word_offset(vaddr_t base, size_t len);
+static uint64_t caprev_shadow_nomap_first_word_mask(vaddr_t base, size_t len);
+static uint64_t caprev_shadow_nomap_last_word_mask(vaddr_t base, size_t len);
+
+/*
+ * Given a capability to an object, derive the offset relative to the base of
+ * the fine-grained bitmap where the first word would be located.
+ */
+
+static ptrdiff_t
+caprev_shadow_nomap_first_word_offset(vaddr_t base)
+{
+	return (ptrdiff_t)(
+	    base / VM_CHERI_REVOKE_GSZ_MEM_NOMAP / 8 / sizeof(uint64_t)) *
+	    sizeof(uint64_t);
+}
+
+static ptrdiff_t
+caprev_shadow_nomap_last_word_offset(vaddr_t base, size_t len)
+{
+	if (len == 0)
+		return caprev_shadow_nomap_first_word_offset(base);
+
+	return (ptrdiff_t)((base + len - 1) / VM_CHERI_REVOKE_GSZ_MEM_NOMAP / 8
+	    / sizeof(uint64_t)) * sizeof(uint64_t);
+}
+
+/*
+ * Derive the mask to be or'd in to the first word to mark the object's
+ * bottom granules as revoked.
+ */
+
+static uint64_t
+caprev_shadow_nomap_first_word_mask(vaddr_t base, size_t len)
+{
+	uint64_t res;
+
+	/* What's the least significant bit's position within the word? */
+	int lsb = (base / VM_CHERI_REVOKE_GSZ_MEM_NOMAP) % (8 *
+	    sizeof(uint64_t));
+
+	if (caprev_shadow_nomap_first_word_offset(base) ==
+	    caprev_shadow_nomap_last_word_offset(base, len)) {
+		/* The object occupies only some bits in the first word */
+
+		int setwidth = len / VM_CHERI_REVOKE_GSZ_MEM_NOMAP;
+
+		if (lsb + setwidth == 64) {
+			/* Object fills this word completely, but does not spill
+			 * to the next */
+
+			res = ~(((uint64_t)1 << lsb) - 1);
+		} else {
+			res = (~(((uint64_t)1 << (lsb + setwidth)) - 1) ^
+			    (~(((uint64_t)1 << lsb) - 1)));
+		}
+	} else {
+		/* The object runs off the end of this word. */
+
+		res = ~(((uint64_t)1 << lsb) - 1);
+	}
+
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+	return __builtin_bswap64(res);
+#elif __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+	return res;
+#else
+#error
+#endif
```

Comment:
Perhaps:

```
    return (le64toh(res));
```

Or the like instead?  (If you are allowed to use <sys/endian.h> here.)